File: frontend/src/shared/api/apiError.ts
import { z } from 'zod'

export interface ApiErrorBody {
  code: string
  messageKey: string
  parameters: Record<string, string | number>
  details: string | null
}

const apiErrorBodySchema: z.ZodType<ApiErrorBody> = z.object({
  code: z.string(),
  messageKey: z.string(),
  parameters: z.record(z.string(), z.union([z.string(), z.number()])),
  details: z.string().nullable(),
})

export function isApiErrorBody(value: unknown): value is ApiErrorBody {
  return apiErrorBodySchema.safeParse(value).success
}

File: backend/GastronomyApp.Contracts/ApiError.cs
namespace GastronomyApp.Contracts;

public sealed record ApiError
{
  public required string Code { get; init; }

  public required string MessageKey { get; init; }

  public required IReadOnlyDictionary<string, string> Parameters { get; init; }

  public string? Details { get; init; }
}
